contracts/MarginCalculator.sol

```diff
@@ -43,19 +45,288 @@ contract MarginCalculator {
+
+        // add new upper bound value for this product at specific time to expiry
+        timeToExpiryValue[productHash][_timeToExpiry] = _value;
+    }
+
+    /**
+     * @notice set spot shock value scaled to 1e27
+     * @dev can only be called by owner
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @param _shockValue spot shock value
+     */
+    function setSpotShock(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut,
+        uint256 _shockValue
+    ) external onlyOwner {
+        require(_shockValue > 0, "MarginCalculator: invalid spot shock value");
+
+        bytes32 productHash = _getProductHash(_underlying, _strike, _collateral, _isPut);
+
+        // set spot shock value in 27 decimals
+        spotShock[productHash] = _shockValue;
+    }
+
+    /**
+     * @notice set oracle deviation (1e27)
+     * @dev can only be called by owner
+     * @param _deviation deviation value
+     */
+    function setOracleDeviation(uint256 _deviation) external onlyOwner {
+        oracleDeviation = _deviation;
+    }
+
+    /**
+     * @notice get dust amount for collateral asset
+     * @param _collateral collateral asset address
+     * @return dust amount (1e27)
+     */
+    function getCollateralDust(address _collateral) external view returns (uint256) {
+        return dust[_collateral];
+    }
+
+    /**
+     * @notice get time to expiry for specific product
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @return array of times to expiry
+     */
+    function getProductTimeToExpiry(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut
+    ) external view returns (uint256[] memory) {
+        bytes32 productHash = _getProductHash(_underlying, _strike, _collateral, _isPut);
+        return productTimeToExpiry[productHash];
+    }
+
+    /**
+     * @notice get option upper bound value for specific time to expiry
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @param _timeToExpiry option time to expiry timestamp
+     * @return option upper bound value (1e27)
+     */
+    function getTimeToExpiryValue(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut,
+        uint256 _timeToExpiry
+    ) external view returns (uint256) {
+        bytes32 productHash = _getProductHash(_underlying, _strike, _collateral, _isPut);
+
+        return timeToExpiryValue[productHash][_timeToExpiry];
```

Comment:
maybe we should have a function that retuns all the values and all of the timeToExpiry via looping through productTimeToExpiry

This is similar to the comment about adding things 1 by 1.  Might be useful to return all for a given product as well as add all for a given product.

Reply:
Why do we need to that in our end ? as we have the function that will return the array of time to expiries, and the function that will return the value for each expiry, so the loop can be done in the side of the dev who need all variables, either in contracts level or fe/web3js as all calls are gasless.

On the other side, I agree with adding a function to set multiple values at once.